pkg/cmd/dump.go

```diff
@@ -0,0 +1,200 @@
+/*
+Licensed to the Apache Software Foundation (ASF) under one or more
+contributor license agreements.  See the NOTICE file distributed with
+this work for additional information regarding copyright ownership.
+The ASF licenses this file to You under the Apache License, Version 2.0
+(the "License"); you may not use this file except in compliance with
+the License.  You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cmd
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"os"
+
+	"github.com/spf13/cobra"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"github.com/apache/camel-k/pkg/client"
+	"github.com/apache/camel-k/pkg/client/camel/clientset/versioned"
+	"github.com/apache/camel-k/pkg/util/kubernetes"
+)
+
+func newCmdDump(rootCmdOptions *RootCmdOptions) (*cobra.Command, *dumpCmdOptions) {
+	options := dumpCmdOptions{
+		RootCmdOptions: rootCmdOptions,
+	}
+	cmd := cobra.Command{
+		Use:     "dump [filename]",
+		Short:   "Dump the state of namespace",
+		Long:    `Dump the state of currently used namespace. If no filename will be specified, the output will be on stdout`,
+		PreRunE: decode(&options),
+		RunE:    options.dump,
+	}
+
+	return &cmd, &options
+}
+
+type dumpCmdOptions struct {
+	*RootCmdOptions
+}
+
+func (o *dumpCmdOptions) dump(_ *cobra.Command, args []string) error {
+	c, err := o.GetCmdClient()
+	if err != nil {
+		return err
+	}
+	if len(args) == 1 {
+		fileName := args[0]
+		writer, err := os.OpenFile(fileName, os.O_RDWR|os.O_CREATE, 0777)
+		if err != nil {
+			return err
+		}
+		dumpNamespace(o.Context, c, o.Namespace, writer)
+		defer writer.Close()
+	} else {
+		dumpNamespace(o.Context, c, o.Namespace, os.Stdout)
+	}
+	return nil
+}
+
+func dumpNamespace(ctx context.Context, c client.Client, ns string, out *os.File) error {
+
+	camelClient, err := versioned.NewForConfig(c.GetConfig())
```

Comment:
I wonder if that could be factorised with the code that dumps the namespace content when e2e tests fail?

Reply:
Do you mean the dump function used in integration tests? I think, it could be done, but it's slightly different. I'll open another issue for this and will work on that.

Reply:
Yes that one. I think it'd be better both share the same "util" dump method as much as possible.

Reply:
Ok, let me open an issue for this

Reply:
https://github.com/apache/camel-k/issues/2172

Reply:
👍🏼